### PR TITLE
Add jade

### PR DESCRIPTION
**Pull request by edouardparis**

### CODE REVIEW

File: gui/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ name = "liana-gui"
 path = "src/main.rs"
 
 [dependencies]
-async-hwi = "0.0.16"
+async-hwi = { git = "https://github.com/wizardsardine/async-hwi", branch = "master" }
```
Comment (edouardparis) [inline gui/Cargo.toml:17]: Waiting feedbacks before releasing new async-hwi version

**Comment by darosior** [comment]
So we still need to use Blockstream Green to setup the device? It looks like [other wallets](https://help.blockstream.com/hc/en-us/articles/9601453403801-Download-a-companion-app-for-Jade) are able to set it up themselves?

> Waiting feedbacks

I would be nice to have a tiny bit of context or documentation about what this PR is supposed to implement or not...

**Comment by pythcoiner** [comment]
> So we still need to use Blockstream Green to setup the device? It looks like [other wallets](https://help.blockstream.com/hc/en-us/articles/9601453403801-Download-a-companion-app-for-Jade) are able to set it up themselves?
> 
> > Waiting feedbacks
> 
> I would be nice to have AT LEAST a tiny bit of context or documentation about what this PR is supposed to implement or not...

There is 3 way:
- blockstream green
- [web usb installer](https://jadefw.blockstream.com/upgrade/fwupgrade.html)
- [python script](https://github.com/Blockstream/Jade/blob/master/FWUPDATE.md) using serial port over usb

**Comment by darosior** [comment]
@pythcoiner what you link is about upgrading the firmware, i'm talking about setting up the device.

**Comment by pythcoiner** [comment]
> @pythcoiner what you link is about upgrading the firmware, i'm talking about setting up the device.

right i misunderstand your last message, i'll factory reset by device as i don't remember if i connect to green first time i spin up this device, and give you feedback

**Comment by darosior** [comment]
Apparently you definitely need a companion app for the first setup.

Anyways @edouardparis i'll need more documentation here. I've set my Jade up but it still only shows "connect to a Green companion app" and isn't detected in the GUI installer.

**Comment by pythcoiner** [comment]
i've done factory reset, import a mnemonic, then get stuck when my jade ask to connect to a compatible device, but at this step if you wait long enough (5-10min) it will leave this screen and (i dont remenber if i set the password before or after) get logged in.
at this step, i can found the jade in the setting menu for register descriptor (trying register `mainnet` descriptor make jade crash ), but the jade is not displayed if i want create a new wallet in `signet` while its detected if i want create a wallet on `mainnet`.

**Comment by pythcoiner** [comment]
Here what i get on the jade while trying to register descriptor from the settings menu:

image

**Comment by pythcoiner** [comment]
I can only unlock my Jade from the installer, not from launched wallet

**Comment by bitcoinhelp** [comment]
> Apparently you definitely need a companion app for the first setup.

@darosior You either need a companion app to initialize Jade with a PIN (but this can be done within Liana for example, you don't need Green) or using the QR PIN Unlock process linked when a user chooses "QR" during setup

> it seems jade is limited in the size of descriptor it can display (3 screen)

Will have our team take a look at this!

**Comment by JamieDriver** [comment]
> it seems jade is limited in the size of descriptor it can display (3 screen)

Right.  You say 'medium size descriptor' ... what lengths of script to do usually use ?
Q: are you using the '@0'... placeholder values to hold the signers and hence shorten the script, or have you got all the xpubs included in the main script string ?

In any case, if we receive a script we deem 'too long to handle' (whatever that is) we ought to error more gracefully than hitting that assert, so def need a fix there.

**Comment by pythcoiner** [comment]
> Q: are you using the `\@0`... placeholder values to hold the signers and hence shorten the script, or have you got all the xpubs included in the main script string ?

i'm no sure, will check, but i think not.....

here an example of a 'big' (minitap)[script](https://github.com/wizardsardine/liana_manual_testing/blob/75fc8c223fcd6da88f97ea09e788bba2ab55cf21/tests/v5/v5-pre-rc1/pyth.md?plain=1#L116)

are you constain by memory on Jade hardware?

**Comment by JamieDriver** [comment]
At the moment we limit scripts on input to 512 byes/chars - if the script gets long how is the user realistically expected to check it on the Jade screen and verify it ?
We allow (and in most cases expect) signers/keys to be passed as `@n` placeholders, which makes the script/policy much easier to read and verify (and the signers/keys can be checked separately).

However - still agree that the assert at display time is a bug and needs fixing.  Cheers for the report.

**Comment by pythcoiner** [comment]
last commit fix the `unlock` issue, but find smth new:
- i first login to my jade w/ its `normal` seed
- get the xpub
- then load a `temporary` seed via QRCode
- want to get the `temporary` xpub for add to my setup

=> still have the xpub related to the `normal` seed, i think we do not fetch the xpub as we link the xpub to the device id?

**Comment by pythcoiner** [comment]
Try something different:
- Restart my jade w/ USB unplugged
- Select `Options`
- Select `Temporary Signer`
- `Continue`
- `Scan QR`
- Scan the temporary seed
- Select `USB`
- Plug USB Cable to the Jade
- Liana still not detect Jade ( but polling)

```
  2024-04-13T06:55:24.554606Z  INFO liana_gui::hw:461: Device("Error { code: -32002, message: Some(\"Cannot process message - hardware locked or uninitialized\"), data: None }")

  2024-04-13T06:55:26.876307Z  INFO liana_gui::hw:461: Device("Error { code: -32002, message: Some(\"Cannot process message - hardware locked or uninitialized\"), data: None }")

  2024-04-13T06:55:29.195030Z  INFO liana_gui::hw:461: Device("Error { code: -32002, message: Some(\"Cannot process message - hardware locked or uninitialized\"), data: None }")

  2024-04-13T06:55:31.516320Z  INFO liana_gui::hw:461: Device("Error { code: -32002, message: Some(\"Cannot process message - hardware locked or uninitialized\"), data: None }")

  2024-04-13T06:55:33.835397Z  INFO liana_gui::hw:461: Device("Error { code: -32002, message: Some(\"Cannot process message - hardware locked or uninitialized\"), data: None }")
```

**Comment by edouardparis** [comment]
After importing a Seed through QrCode and connecting to usb and using https://github.com/Blockstream/Jade
```
 python jade_auth.py 
Fetching jade version info over serial
Connecting...
Sending: {'method': 'get_version_info', 'id': '487301'} as cbor of size 35
Received msg: {'id': '487301', 'result': {'JADE_VERSION': '1.0.30-alpha1', 'JADE_OTA_MAX_CHUNK': 4096, 'JADE_CONFIG': 'NORADIO', 'BOARD_TYPE': 'JADE_V1.1', 'JADE_FEATURES': 'SB', 'IDF_VERSION': 'v5.1.2', 'CHIP_FEATURES': '32000000', 'EFUSEMAC': '782184944810', 'BATTERY_STATUS': 5, 'JADE_STATE': 'LOCKED', 'JADE_NETWORKS': 'ALL', 'JADE_HAS_PIN': True}}
Connected: {'JADE_VERSION': '1.0.30-alpha1', 'JADE_OTA_MAX_CHUNK': 4096, 'JADE_CONFIG': 'NORADIO', 'BOARD_TYPE': 'JADE_V1.1', 'JADE_FEATURES': 'SB', 'IDF_VERSION': 'v5.1.2', 'CHIP_FEATURES': '32000000', 'EFUSEMAC': '782184944810', 'BATTERY_STATUS': 5, 'JADE_STATE': 'LOCKED', 'JADE_NETWORKS': 'ALL', 'JADE_HAS_PIN': True}
Sending: {'method': 'auth_user', 'id': '232825', 'params': {'network': 'mainnet', 'epoch': 1713034159}} as cbor of size 63
Received msg: {'id': '232825', 'error': {'code': -32002, 'message': 'Cannot process message - temporary wallet associated with different connection'}}
Exception causing JadeAPI context exit.
<class 'jadepy.jade_error.JadeError'>
JadeError: -32002 - Cannot process message - temporary wallet associated with different connection (Data: None)
  File "/home/edouard/code/sandbox/Jade/jade_auth.py", line 65, in <module>
    while jade.auth_user(network, http_request_fn) is not True:
  File "/home/edouard/code/sandbox/Jade/jadepy/jade.py", line 743, in auth_user
    return self._jadeRpc('auth_user', params,
  File "/home/edouard/code/sandbox/Jade/jadepy/jade.py", line 325, in _jadeRpc
    result = self._get_result_or_raise_error(reply)
  File "/home/edouard/code/sandbox/Jade/jadepy/jade.py", line 278, in _get_result_or_raise_error
    raise JadeError(e.get('code'), e.get('message'), e.get('data'))
Draining interface...
```

I got the same problem than @pythcoiner with no liana or other app running outside of the .py script

**Comment by pythcoiner** [comment]
i got the same issue w/ Sparrow, definitively not only w/ us:

image

**Comment by JamieDriver** [comment]
The app has to call `auth_user` before it can do anything with wallet keys.  (fyi non-wallet calls like eg. `get_version_info` can work before the `auth_user` call is made).
eg:
connect, call `get_version_info` (if desired), call `add_entropy` (best practice), call `auth_user` (to unlock jade wallet).
- At this point Jade may ask for a PIN and you may need to do an http roundtrip to the blind-oracle to unlock jade - the details are all in the returned data.
- Or, if you have previously done 'Temporary Signer' setup the `auth_user` call should just return true.

The above errors look like `auth_user` has not been called, and the wallet not unlocked for use.

**Comment by JamieDriver** [comment]
Re: `QR Mode` - this is for using airgapped mode, using QRs only.  All (wallet-related) USB or Bluetooth comms will be rejected.

Basically you need to connect to Jade and set/unlock a wallet by some means - USB, Bluetooth or QR code.  Once this has been done the Jade wallet is unlocked *for that method only*.
ie. if you connect and call `auth_user` by USB, and the user unlocks the wallet that way - you can't then send jade `sign_tx` messages over Bluetooth (more importantly, nor an attacker!).

"QR Mode" is for scanning QRs only, all (wallet-related) network comms rejected.

"Temporary Signer" mode is fine - you ought to be able to scan a qr to load the temporary mnemonic/seed, but you must call `auth_user` over the interface you want to use - ie. USB.  In this mode Jade is stateless, in that no wallet data is saved on Jade - the user must enter(or scan) the seed every session.

In "PIN-protected mode", the user sets a wallet phrase and then provides PIN.  An AES key is securely generated by negotiation between the Jade unit and the remote Blind Oracle server.  The wallet root is then encrypted and saved in Jade storage.  ofc the user must enter the correct PIN each session to be able to decrypt the saved wallet.  This method requires a single http roundtrip to the Blind Oracle server.

**Comment by JamieDriver** [comment]
I just tested on master branch the following, and it appeared to work as expected:
* On Jade, Options->Temporary Signer
* `Scan QR` (to load seed data)
* 'USB' for connection mode
* connected to app, which then called `get_version_info` and `auth_user` (no user interaction w/ Jade required), then successfully called 'get_xpub` and `sign_message` (requires user to confirm on Jade before returning signature).

Let me know if this does not appear to be working for you.

**Comment by JamieDriver** [comment]
NOTE: if jade is 'READY' a call to `auth_user` should just return true immediately, pretty much no-op - so if it's easier you can just call this without checking state.  Either path is fine.

(fyi: similarly, before using a descriptor (or multisig) wallet on jade, you need to register the details using `register_descriptor` (or `register_multisig`).  The user must confirm the details of the wallet on Jade, which are then persisted for future use.  At any time the user can delete the registration via the Options menus.
It can be non-trivial for a client app to work out whether the desired wallet details are already/currently registered on Jade - so again calling 'register_descriptor` passing the details of a wallet that is already registered just returns immediately without user interaction, pretty much no-op.  So again it may be easier to call `register_descriptor` before each use, rather than trying to deduce whether that call is strictly necessary.  Again, either course of action is fine.)

**Comment by edouardparis** [comment]
Thank you very much for all the information. It is now I think very clear for me.
I added a doc section about the Jade behavior in https://github.com/wizardsardine/liana/pull/999/commits/8e42e6fc06ae10390c3ff6b1209e062f0f82b55e

**Comment by JamieDriver** [comment]
Please do '@' me if you have any questons / issues.  Really looking forward to seeing Jade and Liana together! :heart:

**Comment by JamieDriver** [comment]
I'll merge the changes we have so far (to Jade master).  
ie. what you have in that `1.0.30-alpha1` build, plus better checking/graceful-error for long descriptors.
(Because we expect the user to manually check the descriptor policy script string and the signers xpubs/details on the Jade UI before agreeing/registering, we sanity-limit the script length and number of signers.  If we need to look at extending these we can do so, but it's more impactful so right now I'm just cleaning up error checking to avoid the assert/abort.  [ ofc also atm Jade does not handle `tr()` ... it's on the list ... ;-) ])

**Comment by pythcoiner** [comment]
> re: [this error](https://github.com/wizardsardine/liana/pull/999#issuecomment-2041074013), can you share the policy/script so I can add it to regression tests ? Cheers.

yep here the descriptor:

``` 
wsh(or_i(and_v(v:thresh(1,pkh([fba81f7f/48'/0'/0']tpubDDsgY25vr3v3ZMeZRMV5GgbRxA2tZXnw96ndrpSHSBqa3h1FQobq2bQnwgCUDsRvAN4jnk8W3BkT5QJGgSfuB3kyM161GkWrLMk1Lkjypus/<4;5>/*),a:pkh([71f42e50/48'/0'/0']tpubDCryh7jnCW3mQTqtSBGk1Pa7CWLVe6ja9q6MMZsrsV3dv1n362xD3vjeZrLtmYMbxH4Rm2P2LLq7YsDqCr5M7m3JbM9BPXFN1iXxHMsh7WN/<6;7>/*)),older(65535)),or_i(and_v(v:thresh(2,pkh([fba81f7f/48'/0'/0']tpubDDsgY25vr3v3ZMeZRMV5GgbRxA2tZXnw96ndrpSHSBqa3h1FQobq2bQnwgCUDsRvAN4jnk8W3BkT5QJGgSfuB3kyM161GkWrLMk1Lkjypus/<2;3>/*),a:pkh([71f42e50/48'/0'/0']tpubDCryh7jnCW3mQTqtSBGk1Pa7CWLVe6ja9q6MMZsrsV3dv1n362xD3vjeZrLtmYMbxH4Rm2P2LLq7YsDqCr5M7m3JbM9BPXFN1iXxHMsh7WN/<4;5>/*)),older(51234)),or_i(and_v(v:thresh(1,pkh([f38d4037/48'/0'/0']tpubDDpE8cYNYT3rYXE1x1dVxYEX3Ud2AyxZVoAM2Dcndwto8Dx9b717oGnPq4N1PtSTBb8AgKjRD9oj1srbYSCmXBAR161q76pEhCiQBmgTSJr/<4;5>/*),a:pkh([71f42e50/48'/0'/0']tpubDCryh7jnCW3mQTqtSBGk1Pa7CWLVe6ja9q6MMZsrsV3dv1n362xD3vjeZrLtmYMbxH4Rm2P2LLq7YsDqCr5M7m3JbM9BPXFN1iXxHMsh7WN/<2;3>/*)),older(41001)),or_d(multi(2,[f38d4037/48'/0'/0']tpubDDpE8cYNYT3rYXE1x1dVxYEX3Ud2AyxZVoAM2Dcndwto8Dx9b717oGnPq4N1PtSTBb8AgKjRD9oj1srbYSCmXBAR161q76pEhCiQBmgTSJr/<0;1>/*,[7eb8fcb3/48'/0'/0']tpubDCxRQvc8EdGgiEKatkGrvCBt7yVtzjJWKNzx9AwHq2DsjjyQZSMAiNd5y6T2xramodXmDPuQUHPySPPfBnMhvYKZSn6Jz4CuS7yg315XYRH/<0;1>/*),and_v(v:thresh(3,pkh([f38d4037/48'/0'/0']tpubDDpE8cYNYT3rYXE1x1dVxYEX3Ud2AyxZVoAM2Dcndwto8Dx9b717oGnPq4N1PtSTBb8AgKjRD9oj1srbYSCmXBAR161q76pEhCiQBmgTSJr/<2;3>/*),a:pkh([7eb8fcb3/48'/0'/0']tpubDCxRQvc8EdGgiEKatkGrvCBt7yVtzjJWKNzx9AwHq2DsjjyQZSMAiNd5y6T2xramodXmDPuQUHPySPPfBnMhvYKZSn6Jz4CuS7yg315XYRH/<2;3>/*),a:pkh([fba81f7f/48'/0'/0']tpubDDsgY25vr3v3ZMeZRMV5GgbRxA2tZXnw96ndrpSHSBqa3h1FQobq2bQnwgCUDsRvAN4jnk8W3BkT5QJGgSfuB3kyM161GkWrLMk1Lkjypus/<0;1>/*),a:pkh([71f42e50/48'/0'/0']tpubDCryh7jnCW3mQTqtSBGk1Pa7CWLVe6ja9q6MMZsrsV3dv1n362xD3vjeZrLtmYMbxH4Rm2P2LLq7YsDqCr5M7m3JbM9BPXFN1iXxHMsh7WN/<0;1>/*)),older(31111)))))))#gcpytqnu
```

**Comment by pythcoiner** [comment]
> How did you manage to load that into Jade ? I get a (graceful) rejection trying to register it, as it's over-long ?

i hit this error testing Liana (not this ONE descriptor, as the descriptor hitting the assert have real coins, i redo the exactly same setup w/ signet tpubs ).
Maybe Liana (or async-hwi) convert it to the wallet policies template?
How do you manage to load in Jade on your side?

**Comment by JamieDriver** [comment]
The verbatim script you gave me was being rejected as too long - once I manually made the keys/xpubs \@ placeholders as you suggest it loaded fine - and in master then gave me the same crash you report above.  :+1: 
I then tested the fixed code (will support up to 5 script display screens btw, up to 512 chars) and added this script as to our unit tests.  
Cheers :rocket: 
(all these fixes will be included in the next fw beta we make - will keep you in the loop)

**Comment by pythcoiner** [comment]
@edouardparis, on async-hwi side do we send the full descriptor or wallet policy?

**Comment by edouardparis** [comment]
[Jade api `register_descriptor`](https://github.com/Blockstream/Jade/blob/master/docs/index.rst#register_descriptor-request) requires that the policy template and keys are send in separated fields. 

From the full descriptor string, async-hwi will extract the template and the keys for example in `register_wallet`: https://github.com/wizardsardine/async-hwi/blob/master/src/jade/mod.rs#L206 with `extract_keys_and_template`: https://github.com/wizardsardine/async-hwi/blob/master/src/utils.rs#L134 

Ex of unit test of `extract_keys_and_template`:
```rust
let res = extract_keys_and_template::<String>("wsh(or_d(multi(2,[b0822927/48'/1'/0'/2']tpubDEvZxV86Br8Knbm9tWcr5Hvmg5cYTYsg92vinqH6Bie6U8ix8CsoN9W11NQygdqVwmHUJpsHXxNsi5gXn36g4xNfLWkMqPuFhRZAmMQ7jjQ/<0;1>/*,[7fc39c07/48'/1'/0'/2']tpubDEvjgXtrUuH3Qtkapny9aE8gN847xiXsf9MDM5XueGf9nrvStqAuBSva3ajGyTvtp8Ti55FvVXsgYSXuS1tQkBeopFuodx2hRUDmQbvKxbZ/<0;1>/*),and_v(v:thresh(2,pkh([b0822927/48'/1'/0'/2']tpubDEvZxV86Br8Knbm9tWcr5Hvmg5cYTYsg92vinqH6Bie6U8ix8CsoN9W11NQygdqVwmHUJpsHXxNsi5gXn36g4xNfLWkMqPuFhRZAmMQ7jjQ/<2;3>/*),a:pkh([7fc39c07/48'/1'/0'/2']tpubDEvjgXtrUuH3Qtkapny9aE8gN847xiXsf9MDM5XueGf9nrvStqAuBSva3ajGyTvtp8Ti55FvVXsgYSXuS1tQkBeopFuodx2hRUDmQbvKxbZ/<2;3>/*),a:pkh([1a1ffd98/48'/1'/0'/2']tpubDFZqzTvGijYb13BC73CkS1er8DrP5YdzMhziN3kWCKUFaW51Yj6ggvf99YpdrkTJy4RT85mxQMHXDiFAKRxzf6BykQgT4pRRBNPshSJJcKo/<0;1>/*)),older(300))))#wp0w3hlw").unwrap();
assert_eq!(res.0, "wsh(or_d(multi(2,@0/<0;1>/*,@1/<0;1>/*),and_v(v:thresh(2,pkh(@0/<2;3>/*),a:pkh(@1/<2;3>/*),a:pkh(@2/<0;1>/*)),older(300))))");
assert_eq!(res.1.len(), 3);
assert_eq!(res.1[0], "[b0822927/48'/1'/0'/2']tpubDEvZxV86Br8Knbm9tWcr5Hvmg5cYTYsg92vinqH6Bie6U8ix8CsoN9W11NQygdqVwmHUJpsHXxNsi5gXn36g4xNfLWkMqPuFhRZAmMQ7jjQ".to_string());
assert_eq!(res.1[1], "[7fc39c07/48'/1'/0'/2']tpubDEvjgXtrUuH3Qtkapny9aE8gN847xiXsf9MDM5XueGf9nrvStqAuBSva3ajGyTvtp8Ti55FvVXsgYSXuS1tQkBeopFuodx2hRUDmQbvKxbZ".to_string());
assert_eq!(res.1[2], "[1a1ffd98/48'/1'/0'/2']tpubDFZqzTvGijYb13BC73CkS1er8DrP5YdzMhziN3kWCKUFaW51Yj6ggvf99YpdrkTJy4RT85mxQMHXDiFAKRxzf6BykQgT4pRRBNPshSJJcKo".to_string());
```

**Review by edouardparis — COMMENTED** [review]
Self-ACK 570f0af35b84372d0e4684f9afd42ea7c900b43e

**Comment by JamieDriver** [comment]
fyi Jade fw `1.0.30` is prepped for release, including all changes/fixes to support Liana, as in the previous alpha build.  
It is available for download/upgrade atm as a 'beta', but that exact build will be promoted to the public latest/stable build in the next 18 hours or so, all being well with final tests.  
(You might want to ensure [or least document] that `1.0.30` is the minimum fw version you need to integrate with Liana ...)

**Pull request merged**